src/DominantJuice.sol

```diff
@@ -0,0 +1,399 @@
+        pledgers.hasBeenRefunded[redeemer] = true;
+
+        // Sending early pledger refund bonus.
+        (bool sendSuccess,) = redeemer.call{value: pledgerRefundBonus}("");
+        require(sendSuccess, "Failed to send refund bonus.");
+
+        emit CycleRefundBonusWithdrawal(redeemer, pledgerRefundBonus);
+    }
+
+    /**
+     * @notice This function is callable after cycleExpiryDate if goal is met. However, for
+     * contingency, where a redeemer is unable to retrieve funds, after a locked period, the
+     * Campaign Manager can withdraw funds for disbursement to remaining redeemers . The unlock
+     * time is hardcoded for two weeks after the cycleExpiryDate to give pledgers confidence
+     * that funds are programmatically guaranteed to remain in the contract for that time.
+     * @param receivingAddress Campaign Manager must call this function but can withdraw funds to
+     * another address if desired.
+     */
+    function creatorWithdraw(address payable receivingAddress, uint256 amount)
+        external
+        payable
+        onlyRole(CAMPAIGN_MANAGER_ROLE)
+    {
+        uint256 unlockTime = campaign.cycleExpiryDate + 1209600; // Two weeks after expiry date
+        require(
+            (hasCycleExpired() && isTargetMet()) || ((block.timestamp > unlockTime) && !isTargetMet()),
+            "Cycle must be expired and successful, or it must be passed the lock period."
+        );
+        if (amount > address(this).balance) revert InsufficientFunds();
+
+        (bool success,) = receivingAddress.call{value: amount}("");
+        require(success, "Failed to withdraw cycle funds.");
+
+        emit CreatorWithdrawal(receivingAddress, amount);
+    }
+
+    /// Getters and Helpers
+
+    function getCampaignInfo() public view returns (Campaign memory) {
+        return campaign;
+    }
+
+    function getBalance() public view returns (uint256) {
+        return address(this).balance;
+    }
+
+    function isTargetMet() public view returns (bool) {
+        if (pledgers.totalAmountPledged >= campaign.cycleTarget) return true;
+        else return false;
+    }
+
+    function hasCycleExpired() public view returns (bool) {
+        if (block.timestamp >= campaign.cycleExpiryDate) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    function getCycleFundingStatus() public view returns (FundingStatus memory) {
+        FundingStatus memory status;
+
+        status.totalPledged = pledgers.totalAmountPledged;
+        status.percentOfGoal = ((100 * pledgers.totalAmountPledged) / campaign.cycleTarget);
+        status.isTargetMet = isTargetMet();
+        status.hasCycleExpired = hasCycleExpired();
+        if (isTargetMet() && address(this).balance == 0) status.hasCreatorWithdrawnAllFunds = true;
+
+        return status;
+    }
+
+    function getPledgerAndTotalWeights(address _pledger) public view returns (uint256, uint256) {
```

Comment:
why is this needed? it's internal processing info that really shouldn't be needed by anyone. If you are using it for tests, I suggest extending the contract with a Test contract to add this functionality.

Reply:
Yes, added getter for tests. Will remove and refactor. Can you very briefly walk me through the high level steps of extending with a Test contract to add the functionality or point me to an example?  I’ve tried to do this before with a different contract, and I couldn’t quite get it to work as intended.

Reply:
Ah, sure.

1. Create a new contract eg `DominantJuiceTestHelper` in another directory (maybe `test`?)
2. Make the contract extend `DominantJuice`
3. Change the fields in `DominantJuice` to `internal` so you can access them from subcontracts
4. Add this getter to `DominantJuiceTestHelper`
5. In tests, deploy `DominantJuiceTestHelper` for the tests that need these additional values